File: Isomorphic Strings.cpp
// Approach: go thru the strings and maintain a hash map for each string. check if already mapped and return false if key doesnt match value, else add the corr char to corr hash map.
// Time: O(n)
// Space: O(1), max 2 hash maps of all ascii chars

class Solution {
public:
    bool isIsomorphic(string s, string t) {
        unordered_map<char,char> sMap;
        unordered_map<char,char> tMap;
        for(int i = 0; i < s.length(); i++) {
            if(sMap.find(s[i]) != sMap.end() && sMap[s[i]] != t[i]) {
                return false;
            }
            sMap[s[i]] = t[i];
            if(tMap.find(t[i]) != tMap.end() && tMap[t[i]] != s[i]) {
                return false;
            }
            tMap[t[i]] = s[i];
        }
        return true;
    }
};

File: isomorphic-strings.py
# We use two hash maps to maintain a one-to-one mapping between characters of s and t, ensuring that each character 
# in s maps uniquely to a character in t and vice versa.
# Time Complexity: O(N), where N is the length of s and t, as we iterate once.
# Space Complexity: O(1), since the hash maps store at most 26 characters (constant space).
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        map_s_t = {}
        map_t_s = {}
        for c1, c2 in zip(s, t):
            if(c1 not in map_s_t) and (c2 not in map_t_s):
                map_s_t[c1] = c2
                map_t_s[c2] = c1
            elif map_s_t.get(c1)!=c2 or map_t_s.get(c2) != c1:
                return False
        return True